File: impl/go/nix/surface.go
package nix

import "fmt"

// The surface: build Nix expressions in Go.
//
// This is what a developer writes instead of learning the Nix language. It
// hands back Expr values, the inspectable core, so everything downstream works
// on data rather than on host closures.
//
// BINDERS ARE HOAS HERE AND NAMED UNDERNEATH.
//
// A lambda is written with a Go function:
//
//	Lam("x", func(x Expr) Expr { return Attrs(P("name", x)) })
//
// which is what makes it composable and frees the caller from inventing names.
// Lam lowers that to a NAMED binder immediately, with a fresh supply, because
// the printer needs a name to print. That two-layer split is the
// recommendation in docs/theory.md section 8, and this file is the only place
// in the package that ever sees a host closure.
//
// The Go surface needs a helper the other three do not: Pair. OCaml and Rust
// have tuple literals and Python has keyword-ordered dicts, so an attribute
// set is written directly as a list of pairs. Go has no tuple literal, so the
// pairs are a named struct and P is its constructor. That is the same shape of
// cost the parent package's JSONValue paid, and again it is ergonomic rather
// than expressive.

// counter is the fresh-name supply.
//
// Ambient package state, like OCaml's and Python's. Rust cannot do this
// without unsafe and uses a thread-local Cell instead, which makes explicit
// what this leaves implicit: the surface is stateful and is not safe to use
// from two goroutines at once.
var counter int

// Reset restarts the fresh-name supply, so a term prints identically twice.
func Reset() { counter = 0 }

func fresh(base string) string {
	counter++
	return fmt.Sprintf("%s%d", base, counter)
}

// Pair is one attribute-set entry. See the note above on why this exists.
type Pair struct {
	Name  string
	Value Expr
}

// P builds a Pair.
func P(name string, value Expr) Pair { return Pair{Name: name, Value: value} }

// Literals

// I is an integer literal.
func I(n int64) Expr { return Int{Value: n} }

// F is a float literal.
func F(f float64) Expr { return Float{Value: f} }

// S is a plain string literal.
func S(s string) Expr { return Str{Parts: []Part{Lit{Text: s}}} }

// Path is a path literal.
func Path(p string) Expr { return PathLit{Text: p} }

// SPath is <nixpkgs>, a search-path lookup.
//
// Impure: it reads NIX_PATH when evaluated, which is why a reproducible caller
// pins it.
func SPath(p string) Expr { return SearchPath{Text: p} }

// V is a variable reference.
func V(x string) Expr { return Var{Name: x} }

// B is true or false, which are ordinary variables in Nix.
func B(b bool) Expr {
	if b {
		return Var{Name: "true"}
	}
	return Var{Name: "false"}
}

// Null is the null literal, likewise an ordinary variable.
func Null() Expr { return Var{Name: "null"} }

// IStr builds an interpolated string.
//
// A string argument is literal text and an Expr is interpolated, so
// IStr("cat ", a, " > $out") is "cat ${a} > $out". Any other type panics,
// which is the price of the variadic: Go cannot express "string or Expr" as a
// parameter type without a sealed interface neither side would want to write
// at every call.
func IStr(pieces ...any) Expr {
	parts := make([]Part, 0, len(pieces))
	for _, p := range pieces {
		switch p := p.(type) {
		case string:
			parts = append(parts, Lit{Text: p})
		case Expr:
			parts = append(parts, Anti{Expr: p})
		default:
			panic(fmt.Sprintf("IStr takes string or Expr, got %T", p))
		}
	}
	return Str{Parts: parts}
}

// Structure

// L is a list literal.
func L(items ...Expr) Expr { return List{Items: items} }

func bindsOf(pairs []Pair) []Binding {
	binds := make([]Binding, 0, len(pairs))
	for _, p := range pairs {
		binds = append(binds, Bind{Path: AttrPath{ID{Name: p.Name}}, Value: p.Value})
	}
	return binds
}

// Attrs is an attribute set. Declaration order is preserved, as Nix source is.
func Attrs(pairs ...Pair) Expr {
	return AttrSet{Recursive: false, Binds: bindsOf(pairs)}
}

// RecAttrs is a recursive attribute set: bindings can refer to each other.
func RecAttrs(pairs ...Pair) Expr {
	return AttrSet{Recursive: true, Binds: bindsOf(pairs)}
}

func idPath(names []string) AttrPath {
	path := make(AttrPath, 0, len(names))
	for _, n := range names {
		path = append(path, ID{Name: n})
	}
	return path
}

// Sel is e.a.b.
func Sel(e Expr, names ...string) Expr {
	return Select{Expr: e, Path: idPath(names)}
}

// SelOr is e.a.b or default.
func SelOr(e Expr, def Expr, names ...string) Expr {
	return Select{Expr: e, Path: idPath(names), Default: def}
}

// App is curried application: App(f, a, b) is ((f a) b).
func App(f Expr, args ...Expr) Expr {
	out := f
	for _, a := range args {
		out = Apply{Func: out, Arg: a}
	}
	return out
}

// Binders

// Lam is a lambda, written with a Go function.
//
// The bound variable is materialised as a fresh NAME, so the result is
// ordinary inspectable syntax rather than a closure.
func Lam(name string, f func(Expr) Expr) Expr {
	n := fresh(name)
	return Lambda{Pattern: PVar{Name: n}, Body: f(Var{Name: n})}
}

// LamAttrs is a lambda taking an attribute set, { a, b ? d }: body.
//
// The body receives a LOOKUP function rather than a struct, because the
// formals are named by the caller at runtime and no host type can give them
// back as fields.
func LamAttrs(formals []Formal, ellipsis bool, f func(func(string) Expr) Expr) Expr {
	body := f(func(name string) Expr { return Var{Name: name} })
	return Lambda{
		Pattern: PSet{Formals: formals, Ellipsis: ellipsis},
		Body:    body,
	}
}

// LetPairs is let a = ...; in body, with the bindings in scope in each other.
func LetPairs(pairs []Pair, body Expr) Expr {
	return Let{Binds: bindsOf(pairs), Body: body}
}

// LetIn binds one value and uses it, without naming it.
//
// The composable form: the caller never sees the generated name, so two
// independently written fragments cannot capture each other's variables.
func LetIn(name string, value Expr, f func(Expr) Expr) Expr {
	n := fresh(name)
	return Let{
		Binds: []Binding{Bind{Path: AttrPath{ID{Name: n}}, Value: value}},
		Body:  f(Var{Name: n}),
	}
}

// Operators

// Plus is a + b.
func Plus(a, b Expr) Expr { return BinOp{Op: OpAdd, Left: a, Right: b} }

// Update is a // b: b wins on a clash.
func Update(a, b Expr) Expr { return BinOp{Op: OpUpdate, Left: a, Right: b} }

// Concat is a ++ b.
func Concat(a, b Expr) Expr { return BinOp{Op: OpConcat, Left: a, Right: b} }

// Eq is a == b.
func Eq(a, b Expr) Expr { return BinOp{Op: OpEq, Left: a, Right: b} }

// Cond is if c then t else f.
func Cond(c, t, f Expr) Expr { return If{Condition: c, Then: t, Else: f} }

// Composability: overlays as mixins

// Overlay is final: prev: { ... }.
//
// Cook and Palsberg's wrapper over a generator (docs/theory.md section 8), and
// Fix is the map out of it.
//
// Overlays form a MONOID under Compose, but only UP TO NIX SEMANTICS, not up
// to syntax. Compose(OverlayID, o) emits ({ } // o) rather than o, and the two
// bracketings of Compose nest // differently; those terms are equal when Nix
// evaluates them and are not equal as syntax. See docs/abstractions.md entry
// 11.
type Overlay func(final, prev Expr) Expr

// OverlayID is the identity overlay: adds nothing.
func OverlayID(_, _ Expr) Expr { return Attrs() }

// Compose applies a first, then b on top, so b wins on a clash.
//
// That matches how a later overlay in a list overrides an earlier one.
func Compose(a, b Overlay) Overlay {
	return func(final, prev Expr) Expr {
		afterA := a(final, prev)
		return Update(afterA, b(final, Update(prev, afterA)))
	}
}

// ComposeAll folds a list of overlays into one.
func ComposeAll(overlays ...Overlay) Overlay {
	out := Overlay(OverlayID)
	for _, o := range overlays {
		out = Compose(out, o)
	}
	return out
}

// Fix closes an overlay into a package set with the knot tied.
//
// Emits (let base = ...; final = base // (overlay final base); in final): the
// fixed point is written OUT in Nix rather than computed here, because the
// point of a transpiler is that the output does the work.
//
// `base` is BOUND, not inlined. `fix` passes it to the overlay as `prev` as
// well as using it on the left of `//`, so inlining it would duplicate the
// whole expression into the output, once per mention. The hand-written Nix a
// reader would compare against binds it too.
func Fix(base Expr, o Overlay) Expr {
	b := fresh("base")
	n := fresh("final")
	added := o(Var{Name: n}, Var{Name: b})
	return Let{
		Binds: []Binding{
			Bind{Path: AttrPath{ID{Name: b}}, Value: base},
			Bind{
				Path:  AttrPath{ID{Name: n}},
				Value: Update(Var{Name: b}, added),
			},
		},
		Body: Var{Name: n},
	}
}

File: impl/python/src/img_drv/nix/surface.py
"""The surface: build Nix expressions in Python.

This is what a developer writes instead of learning the Nix language. It hands
back :mod:`img_drv.nix.ast` values, the inspectable core, so everything
downstream works on data rather than on host closures.

Binders are HOAS here and named underneath
--------------------------------------------

A lambda is written with a Python function::

    lam(lambda x: attrs({"name": x}))

which is what makes it composable and frees the caller from inventing names.
:func:`lam` lowers that to a NAMED binder immediately, with a fresh supply,
because the printer needs a name to print. That two-layer split is the
recommendation in ``docs/theory.md`` section 8, and this module is the only
place in the package that ever sees a host closure.
"""

from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping, Sequence

from . import ast
from .emit import to_nix

__all__ = [
    "apply",
    "attrs",
    "boolean",
    "compose",
    "compose_all",
    "concat",
    "eq",
    "fix",
    "if_",
    "int_",
    "istr",
    "lam",
    "lam_attrs",
    "let_",
    "let_in",
    "listing",
    "null",
    "overlay_id",
    "path",
    "plus",
    "rec_attrs",
    "reset",
    "select",
    "str_",
    "to_nix",
    "update",
    "var",
]

# A fresh-name supply. Names are prefixed so they cannot collide with anything
# a caller wrote, and the counter is resettable rather than global-forever so
# the same term prints the same way twice, which the round-trip law needs.
_counter = [0]


def reset() -> None:
    """Restart the fresh-name supply, so a term prints identically twice."""
    _counter[0] = 0


def _fresh(base: str) -> str:
    _counter[0] += 1
    return f"{base}{_counter[0]}"


# Literals


def int_(n: int) -> ast.Expr:
    return ast.Int(n)


def float_(f: float) -> ast.Expr:
    return ast.Float(f)


def str_(s: str) -> ast.Expr:
    return ast.Str((ast.Lit(s),))


def path(p: str) -> ast.Expr:
    return ast.Path(p)


def var(x: str) -> ast.Expr:
    return ast.Var(x)


def boolean(b: bool) -> ast.Expr:
    return ast.Var("true" if b else "false")


null: ast.Expr = ast.Var("null")


def istr(parts: Iterable[str | ast.Expr]) -> ast.Expr:
    """An interpolated string.

    A ``str`` part is literal text and anything else is interpolated, so
    ``istr(["cat ", a, " > $out"])`` is ``"cat ${a} > $out"``. Distinguishing
    the two by TYPE rather than by a tag is the one place this surface reads
    better than the OCaml one, which needs an explicit `` `S``/`` `E``.
    """
    return ast.Str(
        tuple(ast.Lit(p) if isinstance(p, str) else ast.Anti(p) for p in parts)
    )


# Structure


def listing(items: Iterable[ast.Expr]) -> ast.Expr:
    return ast.List(tuple(items))


def attrs(pairs: Mapping[str, ast.Expr]) -> ast.Expr:
    """An attribute set. Insertion order is preserved, as Nix source is."""
    return ast.AttrSet(
        tuple(ast.Bind((ast.Id(k),), v) for k, v in pairs.items())
    )


def rec_attrs(pairs: Mapping[str, ast.Expr]) -> ast.Expr:
    """A recursive attribute set: bindings can refer to each other by name."""
    return ast.AttrSet(
        tuple(ast.Bind((ast.Id(k),), v) for k, v in pairs.items()),
        recursive=True,
    )


def select(
    e: ast.Expr, names: Sequence[str], default: ast.Expr | None = None
) -> ast.Expr:
    return ast.Select(e, tuple(ast.Id(n) for n in names), default)


def apply(f: ast.Expr, args: Iterable[ast.Expr]) -> ast.Expr:
    """Curried application: ``apply(f, [a, b])`` is ``((f a) b)``."""
    out = f
    for a in args:
        out = ast.Apply(out, a)
    return out


# Binders


def lam(f: Callable[[ast.Expr], ast.Expr], name: str = "x") -> ast.Expr:
    """A lambda, written with a Python function.

    The bound variable is materialised as a fresh NAME, so the result is
    ordinary inspectable syntax rather than a closure.
    """
    n = _fresh(name)
    return ast.Lambda(ast.Pvar(n), f(ast.Var(n)))


def lam_attrs(
    formals: Sequence[tuple[str, ast.Expr | None]],
    f: Callable[[Callable[[str], ast.Expr]], ast.Expr],
    ellipsis: bool = False,
) -> ast.Expr:
    """A lambda taking an attribute set, ``{ a, b ? d }: body``.

    The body receives a LOOKUP function rather than a record, because the
    formals are named by the caller at runtime and no host type can give them
    back as fields.
    """
    body = f(ast.Var)
    return ast.Lambda(ast.Pset(tuple(formals), ellipsis, None), body)


def let_(pairs: Mapping[str, ast.Expr], body: ast.Expr) -> ast.Expr:
    """``let a = ...; in body``, with the bindings in scope in each other."""
    return ast.Let(
        tuple(ast.Bind((ast.Id(k),), v) for k, v in pairs.items()), body
    )


def let_in(
    value: ast.Expr, f: Callable[[ast.Expr], ast.Expr], name: str = "v"
) -> ast.Expr:
    """Bind one value and use it, without naming it.

    The composable form: the caller never sees the generated name, so two
    independently written fragments cannot capture each other's variables.
    """
    n = _fresh(name)
    return ast.Let((ast.Bind((ast.Id(n),), value),), f(ast.Var(n)))


# Operators


def plus(a: ast.Expr, b: ast.Expr) -> ast.Expr:
    return ast.Op(ast.Operator.ADD, a, b)


def update(a: ast.Expr, b: ast.Expr) -> ast.Expr:
    """``a // b``: b wins on a clash."""
    return ast.Op(ast.Operator.UPDATE, a, b)


def concat(a: ast.Expr, b: ast.Expr) -> ast.Expr:
    return ast.Op(ast.Operator.CONCAT, a, b)


def eq(a: ast.Expr, b: ast.Expr) -> ast.Expr:
    return ast.Op(ast.Operator.EQ, a, b)


def if_(c: ast.Expr, t: ast.Expr, f: ast.Expr) -> ast.Expr:
    return ast.If(c, t, f)


# Composability: overlays as mixins
#
# An overlay is `final: prev: { ... }`, which is Cook and Palsberg's wrapper
# over a generator (docs/theory.md section 8). Composition is asymmetric and
# associative with an identity, so overlays form a MONOID, and `fix` is the map
# out of it.

#: final, previous, additions
Overlay = Callable[[ast.Expr, ast.Expr], ast.Expr]


def overlay_id(_final: ast.Expr, _prev: ast.Expr) -> ast.Expr:
    """The identity overlay: adds nothing."""
    return attrs({})


def compose(a: Overlay, b: Overlay) -> Overlay:
    """Apply ``a`` first, then ``b`` on top, so ``b`` wins on a clash.

    That matches how a later overlay in a list overrides an earlier one.
    """

    def composed(final: ast.Expr, prev: ast.Expr) -> ast.Expr:
        after_a = a(final, prev)
        return update(after_a, b(final, update(prev, after_a)))

    return composed


def compose_all(overlays: Iterable[Overlay]) -> Overlay:
    out: Overlay = overlay_id
    for o in overlays:
        out = compose(out, o)
    return out


def fix(base: ast.Expr, o: Overlay) -> ast.Expr:
    """Close an overlay into a package set with the knot tied.

    Emits ``(let final = base // (overlay final base); in final)``: the fixed
    point is written OUT in Nix rather than computed here, because the point of
    a transpiler is that the output does the work.
    """
    n = _fresh("final")
    return ast.Let(
        (ast.Bind((ast.Id(n),), update(base, o(ast.Var(n), base))),), ast.Var(n)
    )
